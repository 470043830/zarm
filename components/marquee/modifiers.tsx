import { DIRECTION_LEFT, DIRECTION_RIGHT, DIRECTION_UP, DIRECTION_DOWN } from './constants';

function getKeyFrameModifier(direction) {
  const dir = direction.toLowerCase();
  switch (dir) {
    case DIRECTION_LEFT:
      return directionLeft;
    case DIRECTION_RIGHT:
      return directionRight;
    case DIRECTION_UP:
      return directionUp;
    case DIRECTION_DOWN:
      return directionDown;
    default:
      return directionLeft;
  }
}

function directionLeft(distance, key) {
  return getHorizontal(-distance, key);
}

function directionRight(distance, key) {
  return getHorizontal(distance, key);
}

function directionUp(distance, key) {
  return getVertical(-distance, key);
}

function directionDown(distance, key) {
  return getVertical(distance, key);
}

function getHorizontal(distance, key) {
  return `@-webkit-keyframes ${key} {
      100% {
        -webkit-transform: translate3d(${distance}px, 0, 0);
        transform: translate3d(${distance}px, 0, 0);
      }
    }
    @keyframes ${key} {
      100% {
        -webkit-transform: translate3d(${distance}px, 0, 0);
        transform: translate3d(${distance}px, 0, 0);
      }
    }`;
}

function getVertical(distance, key) {
  return `@-webkit-keyframes ${key} {
    100% {
      -webkit-transform: translate3d(0, ${distance}px, 0);
      transform: translate3d(0, ${distance}px, 0);
    }
  }
  @keyframes ${key} {
    100% {
      -webkit-transform: translate3d(0, ${distance}px, 0);
      transform: translate3d(0, ${distance}px, 0);
    }
  }`;
}

function animationModifier(duration, loop, delay, key) {
  const infinite = loop ? 'infinite' : 1;
  return `${duration}ms ${key} ${delay}ms linear ${infinite}`;
}
export { getKeyFrameModifier, animationModifier };
