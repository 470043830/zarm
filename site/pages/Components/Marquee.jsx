import Markdown from '@site/components/Markdown';

export default class Page extends Markdown {
  document = () => {
    return {
      document: require('@/components/marquee/demo.md'),
      className: 'marquee-page',
    };
  }
}