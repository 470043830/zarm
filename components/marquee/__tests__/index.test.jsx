import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Marquee from '../index';
import { getKeyFrameModifier, animationModifier } from '../modifiers';

const createMarquee = (props) => {
  return (
    <Marquee {...props}>测试我在滚动</Marquee>
  );
};

describe('Marquee', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <div>
        { createMarquee({id: 1, height: 150, direction: 'right', width: '100%', loop: false, scrollAmount: 6, scrollDelay: 5, prefix: 'zarm-marquee'}) }
      </div>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('component props', () => {
    const upMarquee = mount(createMarquee({id: 1, direction: 'up'}));
    const leftMarquee = mount(createMarquee({id: 1, direction: 'left'}));
    const rightMarquee = mount(createMarquee({id: 2, direction: 'right'}));
    const downMarquee = mount(createMarquee({id: 3, direction: 'down'}));
    const marquee = mount(createMarquee({id: 4, direction: ''}));
    expect(leftMarquee.props().direction).toBe('left');
    expect(upMarquee.props().direction).toBe('up');
    expect(rightMarquee.props().direction).toBe('right');
    expect(downMarquee.props().direction).toBe('down');
    expect(marquee.props().direction).toBe('');
  });

  it('getKeyFrameModifierLeft', () => {
    const fn = getKeyFrameModifier('left');
    const modifier = fn('50', 'zarm-marquee-1').replace(/[\r\n\s]/g,"");
    expect(modifier).toEqual(`@-webkit-keyframeszarm-marquee-1{100%{-webkit-transform:translate3d(-50px,0,0);transform:translate3d(-50px,0,0);}}@keyframeszarm-marquee-1{100%{-webkit-transform:translate3d(-50px,0,0);transform:translate3d(-50px,0,0);}}`);
  });

  it('getKeyFrameModifierRight', () => {
    const fn = getKeyFrameModifier('up');
    const modifier = fn('150', 'zarm-marquee-2').replace(/[\r\n\s]/g,"");
    expect(modifier).toEqual(`@-webkit-keyframeszarm-marquee-2{100%{-webkit-transform:translate3d(0,-150px,0);transform:translate3d(0,-150px,0);}}@keyframeszarm-marquee-2{100%{-webkit-transform:translate3d(0,-150px,0);transform:translate3d(0,-150px,0);}}`);
  });

  it('getKeyFrameModifierRight', () => {
    const fn = getKeyFrameModifier('right');
    const modifier = fn('100', 'zarm-marquee-3').replace(/[\r\n\s]/g,"");
    expect(modifier).toEqual(`@-webkit-keyframeszarm-marquee-3{100%{-webkit-transform:translate3d(100px,0,0);transform:translate3d(100px,0,0);}}@keyframeszarm-marquee-3{100%{-webkit-transform:translate3d(100px,0,0);transform:translate3d(100px,0,0);}}`);
  });

  it('getKeyFrameModifierDown', () => {
    const fn = getKeyFrameModifier('down');
    const modifier = fn('200', 'zarm-marquee-4').replace(/[\r\n\s]/g,"");
    expect(modifier).toEqual(`@-webkit-keyframeszarm-marquee-4{100%{-webkit-transform:translate3d(0,200px,0);transform:translate3d(0,200px,0);}}@keyframeszarm-marquee-4{100%{-webkit-transform:translate3d(0,200px,0);transform:translate3d(0,200px,0);}}`);
  });

  it('getKeyFrameModifierDefault', () => {
    const fn = getKeyFrameModifier('default');
    const modifier = fn('250', 'zarm-marquee-5').replace(/[\r\n\s]/g,"");
    expect(modifier).toEqual(`@-webkit-keyframeszarm-marquee-5{100%{-webkit-transform:translate3d(-250px,0,0);transform:translate3d(-250px,0,0);}}@keyframeszarm-marquee-5{100%{-webkit-transform:translate3d(-250px,0,0);transform:translate3d(-250px,0,0);}}`);
  });
  it('animationModifier', () => {
    const modifier = animationModifier(5, true, 0, 'zarm-marquee-4');
    expect(modifier).toEqual(`5s zarm-marquee-4 0s linear infinite`);
  });
});

