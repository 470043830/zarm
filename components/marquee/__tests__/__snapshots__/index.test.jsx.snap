// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marquee renders correctly 1`] = `
<div>
  <div
    class="za-marquee"
    style="height:150px;width:100%"
  >
    <div
      class="za-marquee-body"
    >
      <div
        class="za-marquee-content"
      >
        测试我在滚动
      </div>
    </div>
  </div>
</div>
`;
